libsigner/src/events.rs

```diff
@@ -0,0 +1,268 @@
+        let srv = TcpListener::bind(listener)?;
+        let bound_addr = srv.local_addr()?;
+        self.sock = Some(srv);
+        self.local_addr = Some(bound_addr.clone());
+        Ok(bound_addr)
+    }
+
+    /// Wait for the node to post something, and then return it.
+    /// Errors are recoverable -- the caller should call this method again even if it returns an
+    /// error.
+    fn next_event(&mut self) -> Result<StackerDBChunksEvent, EventError> {
+        self.with_socket(|event_receiver, server_sock| {
+            let (mut node_sock, _) = server_sock.accept()?;
+
+            // were we asked to terminate?
+            if event_receiver.is_stopped() {
+                return Err(EventError::Terminated);
+            }
+
+            let mut buf = vec![];
+            node_sock.read_to_end(&mut buf)?;
+
+            let (verb, path, headers, body_offset) = decode_http_request(&buf)?;
+            if verb != "POST" {
+                return Err(EventError::MalformedRequest(format!(
+                    "Unrecognized verb '{}'",
+                    &verb
+                )));
+            }
+            if path != "/stackerdb_chunks" {
+                return Err(EventError::UnrecognizedEvent(path));
+            }
+
+            let body = decode_http_body(&headers, &buf[body_offset..])?;
+            let event: StackerDBChunksEvent = serde_json::from_slice(&body).map_err(|e| {
+                EventError::Deserialize(format!("Could not decode body to JSON: {:?}", &e))
+            })?;
+            Ok(event)
+        })?
+    }
+
+    /// Determine if the receiver is hung up
+    fn is_stopped(&self) -> bool {
+        self.stop_signal.load(Ordering::SeqCst)
+    }
+
+    /// Forward an event
+    /// Return true on success; false on error.
+    /// Returning false terminates the event receiver.
+    fn forward_event(&mut self, ev: StackerDBChunksEvent) -> bool {
+        if self.out_channels.len() == 0 {
+            // nothing to do
+            error!("No channels connected to event receiver");
+            return false;
+        } else if self.out_channels.len() == 1 {
+            // avoid a clone
+            if let Err(e) = self.out_channels[0].send(ev) {
+                error!("Failed to send to signer runloop: {:?}", &e);
+                return false;
+            }
+            return true;
+        } else {
+            for (i, out_channel) in self.out_channels.iter().enumerate() {
+                if let Err(e) = out_channel.send(ev.clone()) {
+                    error!("Failed to send to signer runloop #{}: {:?}", i, &e);
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    /// Add an event consumer.  A received event will be forwarded to this Sender.
+    fn add_consumer(&mut self, out_channel: Sender<StackerDBChunksEvent>) {
+        self.out_channels.push(out_channel);
+    }
```

Comment:
Is there any plan to include a mechanism for new consumers to sync up with old events? Or do we assume consumers have to separately query the node RPC endpoint to get the latest StackerDB state when starting to consume events?

Reply:
Not sure, but I would anticipate there being some initial state per signing round the signer should catch up on when connecting - such as knowing which the other signers are. I don't have the full picture in my head though so you know this better than I.